In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [5]:
df = pd.read_csv('/Users/atharvadeo/Downloads/archive/Combined_Flights_2018.csv')

In [6]:
pd.set_option('display.max_columns', 500)

In [7]:
df.head()

FlightDate            Airline Origin Dest  Cancelled  Diverted  CRSDepTime  \
0  2018-01-23  Endeavor Air Inc.    ABY  ATL      False     False        1202   
1  2018-01-24  Endeavor Air Inc.    ABY  ATL      False     False        1202   
2  2018-01-25  Endeavor Air Inc.    ABY  ATL      False     False        1202   
3  2018-01-26  Endeavor Air Inc.    ABY  ATL      False     False        1202   
4  2018-01-27  Endeavor Air Inc.    ABY  ATL      False     False        1400   

   DepTime  DepDelayMinutes  DepDelay  ArrTime  ArrDelayMinutes  AirTime  \
0   1157.0              0.0      -5.0   1256.0              0.0     38.0   
1   1157.0              0.0      -5.0   1258.0              0.0     36.0   
2   1153.0              0.0      -9.0   1302.0              0.0     40.0   
3   1150.0              0.0     -12.0   1253.0              0.0     35.0   
4   1355.0              0.0      -5.0   1459.0              0.0     36.0   

   CRSElapsedTime  ActualElapsedTime  Distance  Year  Quarter  Month  \
0            62.0               59.0     145.0  2018        1      1   
1            62.0               61.0     145.0  2018        1      1   
2            62.0               69.0     145.0  2018        1      1   
3            62.0               63.0     145.0  2018        1      1   
4            60.0               64.0     145.0  2018        1      1   

   DayofMonth  DayOfWeek Marketing_Airline_Network  \
0          23          2                        DL   
1          24          3                        DL   
2          25          4                        DL   
3          26          5                        DL   
4          27          6                        DL   

  Operated_or_Branded_Code_Share_Partners  DOT_ID_Marketing_Airline  \
0                            DL_CODESHARE                     19790   
1                            DL_CODESHARE                     19790   
2                            DL_CODESHARE                     19790   
3                            DL_CODESHARE                     19790   
4                            DL_CODESHARE                     19790   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          DL                             3298   
1                          DL                             3298   
2                          DL                             3298   
3                          DL                             3298   
4                          DL                             3298   

  Operating_Airline  DOT_ID_Operating_Airline IATA_Code_Operating_Airline  \
0                9E                     20363                          9E   
1                9E                     20363                          9E   
2                9E                     20363                          9E   
3                9E                     20363                          9E   
4                9E                     20363                          9E   

  Tail_Number  Flight_Number_Operating_Airline  OriginAirportID  \
0      N8928A                             3298            10146   
1      N800AY                             3298            10146   
2      N8836A                             3298            10146   
3      N800AY                             3298            10146   
4      N8839E                             3298            10146   

   OriginAirportSeqID  OriginCityMarketID OriginCityName OriginState  \
0             1014602               30146     Albany, GA          GA   
1             1014602               30146     Albany, GA          GA   
2             1014602               30146     Albany, GA          GA   
3             1014602               30146     Albany, GA          GA   
4             1014602               30146     Albany, GA          GA   

   OriginStateFips OriginStateName  OriginWac  DestAirportID  \
0               13         Georgia         34          10397   
1               13         Georgia         34     

In [8]:
df.tail()

FlightDate                      Airline Origin Dest  Cancelled  \
5689507  2018-09-11  Air Wisconsin Airlines Corp    SCE  IAD      False   
5689508  2018-09-11  Air Wisconsin Airlines Corp    IAD  GSO      False   
5689509  2018-09-11  Air Wisconsin Airlines Corp    EVV  ORD      False   
5689510  2018-09-11  Air Wisconsin Airlines Corp    ORD  HPN      False   
5689511  2018-09-11  Air Wisconsin Airlines Corp    HPN  ORD      False   

         Diverted  CRSDepTime  DepTime  DepDelayMinutes  DepDelay  ArrTime  \
5689507     False        1445   1433.0              0.0     -12.0   1515.0   
5689508     False        1235   1224.0              0.0     -11.0   1344.0   
5689509     False        1030   1016.0              0.0     -14.0   1137.0   
5689510     False        1410   1403.0              0.0      -7.0   1717.0   
5689511     False        1800   1754.0              0.0      -6.0   1919.0   

         ArrDelayMinutes  AirTime  CRSElapsedTime  ActualElapsedTime  \
5689507              0.0     35.0            61.0               42.0   
5689508              0.0     43.0            80.0               80.0   
5689509              0.0     54.0            94.0               81.0   
5689510              0.0    104.0           136.0              134.0   
5689511              0.0    116.0           153.0              145.0   

         Distance  Year  Quarter  Month  DayofMonth  DayOfWeek  \
5689507     133.0  2018        3      9          11          2   
5689508     239.0  2018        3      9          11          2   
5689509     272.0  2018        3      9          11          2   
5689510     738.0  2018        3      9          11          2   
5689511     738.0  2018        3      9          11          2   

        Marketing_Airline_Network Operated_or_Branded_Code_Share_Partners  \
5689507                        UA                            UA_CODESHARE   
5689508                        UA                            UA_CODESHARE   
5689509                        UA                            UA_CODESHARE   
5689510                        UA                            UA_CODESHARE   
5689511                        UA                            UA_CODESHARE   

         DOT_ID_Marketing_Airline IATA_Code_Marketing_Airline  \
5689507                     19977                          UA   
5689508                     19977                          UA   
5689509                     19977                          UA   
5689510                     19977                          UA   
5689511                     19977                          UA   

         Flight_Number_Marketing_Airline Operating_Airline  \
5689507                             3839                ZW   
5689508                             3838                ZW   
5689509                             3837                ZW   
5689510                             3837                ZW   
5689511                             3836                ZW   

         DOT_ID_Operating_Airline IATA_Code_Operating_Airline Tail_Number  \
5689507                     20046                          ZW      N437AW   
5689508                     20046                          ZW      N410AW   
5689509                     20046                          ZW      N467AW   
5689510                     20046                          ZW      N440AW   
5689511                     20046                          ZW      N440AW   

         Flight_Number_Operating_Airline  OriginAirportID  OriginAirportSeqID  \
5689507                             3839            14711             1471105   
5689508                             3838            12264             1226402   
5689509                             3837            11612             1161206   
5689510                             3837            13930             1393006   
5689511                             3836            12197             1219702   

         OriginCityMarketID     OriginCityName OriginState  OriginStateFips  \
568

In [9]:
df.info(verbose=False)
#use verbose for a short answer

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5689512 entries, 0 to 5689511
Columns: 61 entries, FlightDate to DivAirportLandings
dtypes: bool(2), float64(19), int64(22), object(18)
memory usage: 2.5+ GB


The df.sample() method is used to select a random sample of rows from your DataFrame. This is helpful when you want to quickly look at a representative subset of your data, especially if your DataFrame is very large. You can specify the number of rows to sample or the fraction of rows to sample.

In [10]:
df.sample(frac=0.1, random_state=42)

FlightDate                 Airline Origin Dest  Cancelled  Diverted  \
2846883  2018-03-24  Hawaiian Airlines Inc.    LIH  HNL      False     False   
3173911  2018-04-23       Endeavor Air Inc.    BOS  PIT      False     False   
2435108  2018-02-16  Southwest Airlines Co.    HOU  ECP      False     False   
2234995  2018-12-07  Southwest Airlines Co.    MDW  MCO      False     False   
3986585  2018-06-03  Southwest Airlines Co.    LAS  LIT      False     False   
...             ...                     ...    ...  ...        ...       ...   
1414395  2018-11-09             Comair Inc.    JAN  CLT      False     False   
1512299  2018-11-28    Delta Air Lines Inc.    MSP  ORD      False     False   
1989567  2018-12-10       Endeavor Air Inc.    IND  RDU       True     False   
2759150  2018-03-17   United Air Lines Inc.    FLL  EWR      False     False   
2833668  2018-03-15       Republic Airlines    ORD  DTW      False     False   

         CRSDepTime  DepTime  DepDelayMinutes  DepDelay  ArrTime  \
2846883        1900   1855.0              0.0      -5.0   1927.0   
3173911        2055   2111.0             16.0      16.0   2316.0   
2435108        1650   1728.0             38.0      38.0   1903.0   
2234995        1845   1851.0              6.0       6.0   2214.0   
3986585        1200   1201.0              1.0       1.0   1639.0   
...             ...      ...              ...       ...      ...   
1414395        1428   1434.0              6.0       6.0   1719.0   
1512299         650    646.0              0.0      -4.0    808.0   
1989567        1310      NaN              NaN       NaN      NaN   
2759150        1423   1418.0              0.0      -5.0   1726.0   
2833668        1545   1544.0              0.0      -1.0   1755.0   

         ArrDelayMinutes  AirTime  CRSElapsedTime  ActualElapsedTime  \
2846883              0.0     19.0            35.0               32.0   
3173911             27.0     80.0           114.0              125.0   
2435108             43.0     75.0            90.0               95.0   
2234995              0.0    127.0           150.0              143.0   
3986585              0.0    149.0           175.0              158.0   
...                  ...      ...             ...                ...   
1414395              0.0     75.0           113.0              105.0   
1512299              0.0     56.0           106.0               82.0   
1989567              NaN      NaN           105.0                NaN   
2759150              0.0    154.0           188.0              188.0   
2833668              0.0     47.0            82.0               71.0   

         Distance  Year  Quarter  Month  DayofMonth  DayOfWeek  \
2846883     102.0  2018        1      3          24          6   
3173911     496.0  2018        2      4          23          1   
2435108     571.0  2018        1      2          16          5   
2234995     990.0  2018        4     12           7          5   
3986585    1294.0  2018        2      6           3          7   
...           ...   ...      ...    ...         ...        ...   
1414395     562.0  2018        4     11           9          5   
1512299     334.0  2018        4     11          28          3   
1989567     489.0  2018        4     12          10          1   
2759150    1065.0  2018        1      3          17          6   
2833668     235.0  2018        1      3          15          4   

        Marketing_Airline_Network Operated_or_Branded_Code_Share_Partners  \
2846883                        HA                                      HA   
3173911                        DL                            DL_CODESHARE   
2435108                        WN                                      WN   
2234995                        WN                                      WN   
3986585                        WN                                      WN   
...                           ...                                     ...   
1414395                        AA 

df.columns is an attribute of the DataFrame, not a method. Attributes are characteristics or properties of an object, while methods are actions that can be performed on an object. In this case, df.columns directly accesses the index object containing the column labels, so you don't need the parentheses.

In [11]:
df.columns

Index(['FlightDate', 'Airline', 'Origin', 'Dest', 'Cancelled', 'Diverted',
       'CRSDepTime', 'DepTime', 'DepDelayMinutes', 'DepDelay', 'ArrTime',
       'ArrDelayMinutes', 'AirTime', 'CRSElapsedTime', 'ActualElapsedTime',
       'Distance', 'Year', 'Quarter', 'Month', 'DayofMonth', 'DayOfWeek',
       'Marketing_Airline_Network', 'Operated_or_Branded_Code_Share_Partners',
       'DOT_ID_Marketing_Airline', 'IATA_Code_Marketing_Airline',
       'Flight_Number_Marketing_Airline', 'Operating_Airline',
       'DOT_ID_Operating_Airline', 'IATA_Code_Operating_Airline',
       'Tail_Number', 'Flight_Number_Operating_Airline', 'OriginAirportID',
       'OriginAirportSeqID', 'OriginCityMarketID', 'OriginCityName',
       'OriginState', 'OriginStateFips', 'OriginStateName', 'OriginWac',
       'DestAirportID', 'DestAirportSeqID', 'DestCityMarketID', 'DestCityName',
       'DestState', 'DestStateFips', 'DestStateName', 'DestWac', 'DepDel15',
       'DepartureDelayGroups', 'DepTimeBlk', 'TaxiOu

df.describe() is a very useful method for getting a quick summary of the numerical columns in your DataFrame. It calculates and displays various descriptive statistics such as:

count: The number of non-null values.

mean: The average value.

std: The standard deviation, which measures the spread of the data.

min: The minimum value.

25%: The first quartile (25th percentile).

50%: The median (50th percentile).

75%: The third quartile (75th percentile).

max: The maximum value.

In [12]:
df.describe()

CRSDepTime       DepTime  DepDelayMinutes      DepDelay  \
count  5.689512e+06  5.604132e+06     5.602937e+06  5.602937e+06   
mean   1.330341e+03  1.334473e+03     1.346201e+01  1.014888e+01   
std    4.913570e+02  5.048563e+02     4.420411e+01  4.536897e+01   
min    1.000000e+00  1.000000e+00     0.000000e+00 -1.280000e+03   
25%    9.150000e+02  9.160000e+02     0.000000e+00 -6.000000e+00   
50%    1.323000e+03  1.327000e+03     0.000000e+00 -2.000000e+00   
75%    1.735000e+03  1.745000e+03     7.000000e+00  7.000000e+00   
max    2.359000e+03  2.400000e+03     2.625000e+03  2.625000e+03   

            ArrTime  ArrDelayMinutes       AirTime  CRSElapsedTime  \
count  5.598897e+06     5.586619e+06  5.580241e+06    5.689506e+06   
mean   1.462641e+03     1.368363e+01  1.101716e+02    1.389089e+02   
std    5.405804e+02     4.402848e+01  7.058990e+01    7.288408e+01   
min    1.000000e+00     0.000000e+00 -1.244000e+03   -9.900000e+01   
25%    1.046000e+03     0.000000e+00  5.900000e+01    8.600000e+01   
50%    1.503000e+03     0.000000e+00  9.200000e+01    1.200000e+02   
75%    1.915000e+03     8.000000e+00  1.390000e+02    1.700000e+02   
max    2.400000e+03     2.635000e+03  6.960000e+02    1.645000e+03   

       ActualElapsedTime      Distance       Year       Quarter         Month  \
count       5.587165e+06  5.689512e+06  5689512.0  5.689512e+06  5.689512e+06   
mean        1.344377e+02  7.878012e+02     2018.0  2.765149e+00  7.290665e+00   
std         7.258416e+01  5.943094e+02        0.0  1.115764e+00  3.503684e+00   
min        -1.228000e+03  1.600000e+01     2018.0  1.000000e+00  1.000000e+00   
25%         8.200000e+01  3.490000e+02     2018.0  2.000000e+00  4.000000e+00   
50%         1.170000e+02  6.280000e+02     2018.0  3.000000e+00  8.000000e+00   
75%         1.650000e+02  1.024000e+03     2018.0  4.000000e+00  1.000000e+01   
max         7.570000e+02  4.983000e+03     2018.0  4.000000e+00  1.200000e+01   

         DayofMonth     DayOfWeek  DOT_ID_Marketing_Airline  \
count  5.689512e+06  5.689512e+06              5.689512e+06   
mean   1.577653e+01  3.942995e+00              1.983425e+04   
std    8.778083e+00  2.001097e+00              3.149599e+02   
min    1.000000e+00  1.000000e+00              1.939300e+04   
25%    8.000000e+00  2.000000e+00              1.969000e+04   
50%    1.600000e+01  4.000000e+00              1.980500e+04   
75%    2.300000e+01  6.000000e+00              1.997700e+04   
max    3.100000e+01  7.000000e+00              2.117100e+04   

       Flight_Number_Marketing_Airline  DOT_ID_Operating_Airline  \
count                     5.689512e+06              5.689512e+06   
mean                      2.638920e+03              1.998161e+04   
std                       1.914474e+03              4.149514e+02   
min                       1.000000e+00              1.939300e+04   
25%                       9.420000e+02              1.968700e+04   
50%                       2.127000e+03              1.997700e+04   
75%                       4.285000e+03              2.036800e+04   
max                       9.385000e+03              2.117100e+04   

       Flight_Number_Operating_Airline  OriginAirportID  OriginAirportSeqID  \
count                     5.689512e+06     5.689512e+06        5.689512e+06   
mean                      2.638847e+03     1.271578e+04        1.271582e+06   
std                       1.914574e+03     1.505246e+03        1.505245e+05   
min                       1.000000e+00     1.013500e+04        1.013505e+06   
25%                       9.420000e+02     1.129200e+04        1.129202e+06   
50%                       2.127000e+03     1.288900e+04        1.288903e+06   
75%                       4.285000e+03     1.405700e+04        1.405702e+06   
max                       9.385000e+03     1.621800e+04        1.621802e+06   

       OriginCityMarketID  OriginStateFips     OriginWac  DestAirportID  \
count        5.689512e+06     5.689512e+06  5.689512e+06   5.689512

In [13]:
df.columns

Index(['FlightDate', 'Airline', 'Origin', 'Dest', 'Cancelled', 'Diverted',
       'CRSDepTime', 'DepTime', 'DepDelayMinutes', 'DepDelay', 'ArrTime',
       'ArrDelayMinutes', 'AirTime', 'CRSElapsedTime', 'ActualElapsedTime',
       'Distance', 'Year', 'Quarter', 'Month', 'DayofMonth', 'DayOfWeek',
       'Marketing_Airline_Network', 'Operated_or_Branded_Code_Share_Partners',
       'DOT_ID_Marketing_Airline', 'IATA_Code_Marketing_Airline',
       'Flight_Number_Marketing_Airline', 'Operating_Airline',
       'DOT_ID_Operating_Airline', 'IATA_Code_Operating_Airline',
       'Tail_Number', 'Flight_Number_Operating_Airline', 'OriginAirportID',
       'OriginAirportSeqID', 'OriginCityMarketID', 'OriginCityName',
       'OriginState', 'OriginStateFips', 'OriginStateName', 'OriginWac',
       'DestAirportID', 'DestAirportSeqID', 'DestCityMarketID', 'DestCityName',
       'DestState', 'DestStateFips', 'DestStateName', 'DestWac', 'DepDel15',
       'DepartureDelayGroups', 'DepTimeBlk', 'TaxiOu

In [14]:
df[['Airline']].describe()

Airline
count                  5689512
unique                      28
top     Southwest Airlines Co.
freq                   1352552

In [15]:
df.shape
#Returns a tuple that gives the dimensions of DataFrame


(5689512, 61)

Subsetting a DataFrame

In [16]:
df[df.columns[:5]]  #uses the 5 columns and prints them,

FlightDate                      Airline Origin Dest  Cancelled
0        2018-01-23            Endeavor Air Inc.    ABY  ATL      False
1        2018-01-24            Endeavor Air Inc.    ABY  ATL      False
2        2018-01-25            Endeavor Air Inc.    ABY  ATL      False
3        2018-01-26            Endeavor Air Inc.    ABY  ATL      False
4        2018-01-27            Endeavor Air Inc.    ABY  ATL      False
...             ...                          ...    ...  ...        ...
5689507  2018-09-11  Air Wisconsin Airlines Corp    SCE  IAD      False
5689508  2018-09-11  Air Wisconsin Airlines Corp    IAD  GSO      False
5689509  2018-09-11  Air Wisconsin Airlines Corp    EVV  ORD      False
5689510  2018-09-11  Air Wisconsin Airlines Corp    ORD  HPN      False
5689511  2018-09-11  Air Wisconsin Airlines Corp    HPN  ORD      False

[5689512 rows x 5 columns]

In [17]:
df.iloc[5]
#It returns a pandas Series containing all the column values for that specific row.

FlightDate                   2018-01-28
Airline               Endeavor Air Inc.
Origin                              ABY
Dest                                ATL
Cancelled                         False
                            ...        
ArrDel15                            1.0
ArrivalDelayGroups                  1.0
ArrTimeBlk                    1300-1359
DistanceGroup                         1
DivAirportLandings                  0.0
Name: 5, Length: 61, dtype: object

In [18]:
df.iloc[:, [1,2]]
#The : before the comma means "select all rows".
#[1, 2] after the comma means "select the columns at integer positions 1 and 2".

Airline Origin
0                  Endeavor Air Inc.    ABY
1                  Endeavor Air Inc.    ABY
2                  Endeavor Air Inc.    ABY
3                  Endeavor Air Inc.    ABY
4                  Endeavor Air Inc.    ABY
...                              ...    ...
5689507  Air Wisconsin Airlines Corp    SCE
5689508  Air Wisconsin Airlines Corp    IAD
5689509  Air Wisconsin Airlines Corp    EVV
5689510  Air Wisconsin Airlines Corp    ORD
5689511  Air Wisconsin Airlines Corp    HPN

[5689512 rows x 2 columns]

In [19]:
df.query('(DepTime > 1130) and (Origin == "DRO")')

FlightDate                Airline Origin Dest  Cancelled  Diverted  \
24675    2018-01-19  SkyWest Airlines Inc.    DRO  DEN      False     False   
24720    2018-01-19  SkyWest Airlines Inc.    DRO  DEN      False     False   
24991    2018-01-19  SkyWest Airlines Inc.    DRO  DEN      False     False   
27338    2018-01-05  SkyWest Airlines Inc.    DRO  DEN      False     False   
27380    2018-01-05  SkyWest Airlines Inc.    DRO  DEN      False     False   
...             ...                    ...    ...  ...        ...       ...   
5576606  2018-09-14  SkyWest Airlines Inc.    DRO  DEN      False     False   
5578380  2018-09-15  SkyWest Airlines Inc.    DRO  DEN      False     False   
5689157  2018-09-11  Trans States Airlines    DRO  DEN      False     False   
5689165  2018-09-11  Trans States Airlines    DRO  DEN      False     False   
5689258  2018-09-11  Trans States Airlines    DRO  DEN      False     False   

         CRSDepTime  DepTime  DepDelayMinutes  DepDelay  ArrTime  \
24675          1805   1742.0              0.0     -23.0   1912.0   
24720          1545   1529.0              0.0     -16.0   1634.0   
24991          1330   1321.0              0.0      -9.0   1427.0   
27338          1805   1747.0              0.0     -18.0   1857.0   
27380          1550   1539.0              0.0     -11.0   1641.0   
...             ...      ...              ...       ...      ...   
5576606        1551   1545.0              0.0      -6.0   1651.0   
5578380        1551   1550.0              0.0      -1.0   1659.0   
5689157        1551   1543.0              0.0      -8.0   1650.0   
5689165        1750   1745.0              0.0      -5.0   1858.0   
5689258        1310   1302.0              0.0      -8.0   1400.0   

         ArrDelayMinutes  AirTime  CRSElapsedTime  ActualElapsedTime  \
24675                0.0     44.0            77.0               90.0   
24720                0.0     41.0            70.0               65.0   
24991                0.0     46.0            69.0               66.0   
27338                0.0     44.0            75.0               70.0   
27380                0.0     45.0            69.0               62.0   
...                  ...      ...             ...                ...   
5576606              0.0     47.0            69.0               66.0   
5578380              0.0     47.0            69.0               69.0   
5689157              0.0     49.0            69.0               67.0   
5689165              0.0     50.0            76.0               73.0   
5689258              0.0     46.0            70.0               58.0   

         Distance  Year  Quarter  Month  DayofMonth  DayOfWeek  \
24675       250.0  2018        1      1          19          5   
24720       250.0  2018        1      1          19          5   
24991       250.0  2018        1      1          19          5   
27338       250.0  2018        1      1           5          5   
27380       250.0  2018        1      1           5          5   
...           ...   ...      ...    ...         ...        ...   
5576606     250.0  2018        3      9          14          5   
5578380     250.0  2018        3      9          15          6   
5689157     250.0  2018        3      9          11          2   
5689165     250.0  2018        3      9          11          2   
5689258     250.0  2018        3      9          11          2   

        Marketing_Airline_Network Operated_or_Branded_Code_Share_Partners  \
24675                          UA                            UA_CODESHARE   
24720                          UA                            UA_CODESHARE   
24991                          UA                            UA_CODESHARE   
27338                          UA                            UA_CODESHARE   
27380                          UA                            UA_CODESHARE   
...                           ...                                     ...   
5576606                        UA             

.quantile([0.25, 0.75]) calculates the values at the specified quantiles (percentiles). In this case, 0.25 represents the 25th percentile (the first quartile) and 0.75 represents the 75th percentile (the third quartile).

In [20]:
df['DepTime'].quantile([0.25, 0.75])

0.25     916.0
0.75    1745.0
Name: DepTime, dtype: float64

df.groupby('Airline') groups the DataFrame by the unique values in the 'Airline' column.
[['DepDelay']] is then applied to the grouped object. This selects the 'DepDelay' column specifically as a DataFrame. This is important because the .agg() method is being applied to this selection.

In [21]:
df.groupby('Airline')[['DepDelay','ArrDelay']].agg(['mean', 'min', 'max'])

DepDelay                  \
                                                mean     min     max   
Airline                                                                
Air Wisconsin Airlines Corp                11.405747   -49.0  1296.0   
Alaska Airlines Inc.                        2.021924  -114.0   839.0   
Allegiant Air                              12.922575   -53.0  1462.0   
American Airlines Inc.                     10.121187   -31.0  2109.0   
Cape Air                                   -2.729355   -50.0   430.0   
Capital Cargo International                10.441951   -45.0   841.0   
Comair Inc.                                 9.471686   -34.0  1121.0   
Commutair Aka Champlain Enterprises, Inc.  24.726451   -31.0  1352.0   
Compass Airlines                           10.080122 -1280.0  2625.0   
Delta Air Lines Inc.                        5.690527   -40.0  1207.0   
Empire Airlines Inc.                        8.480185   -42.0   655.0   
Endeavor Air Inc.                           9.644629   -38.0  1921.0   
Envoy Air                                   6.885426   -40.0  1163.0   
ExpressJet Airlines Inc.                   11.855062   -36.0  1522.0   
Frontier Airlines Inc.                     19.684102   -60.0  1254.0   
GoJet Airlines, LLC d/b/a United Express   11.872817  -611.0  1545.0   
Hawaiian Airlines Inc.                      0.913099   -37.0  2482.0   
Horizon Air                                 3.073117   -54.0   640.0   
JetBlue Airways                            15.869267   -32.0  1489.0   
Mesa Airlines Inc.                         11.612683   -30.0  1789.0   
Peninsula Airways Inc.                     10.671280   -33.0   470.0   
Republic Airlines                           7.699234   -48.0  1270.0   
SkyWest Airlines Inc.                       9.949630   -60.0  2098.0   
Southwest Airlines Co.                     11.138968   -49.0   713.0   
Spirit Air Lines                            9.868512   -47.0  1527.0   
Trans States Airlines                      16.788899   -51.0  1519.0   
United Air Lines Inc.                      10.060413  -122.0  1431.0   
Virgin America                              6.333836   -24.0   520.0   

                                            ArrDelay                  
                                                mean     min     max  
Airline                                                               
Air Wisconsin Airlines Corp                 8.013017   -70.0  1292.0  
Alaska Airlines Inc.                       -1.180165  -106.0   842.0  
Allegiant Air                               9.983518   -98.0  1505.0  
American Airlines Inc.                      6.514280   -70.0  2153.0  
Cape Air                                   -1.413897   -45.0   446.0  
Capital Cargo International                 6.588072   -51.0   814.0  
Comair Inc.                                 4.860667   -56.0  1110.0  
Commutair Aka Champlain Enterprises, Inc.  21.579446   -60.0  1353.0  
Compass Airlines                            6.160631 -1290.0  2635.0  
Delta Air Lines Inc.                       -2.225013   -83.0  1206.0  
Empire Airlines Inc.                        8.944804   -48.0   654.0  
Endeavor Air Inc.                           3.454431   -83.0  1916.0  
Envoy Air                                   4.502394   -61.0  1152.0  
ExpressJet Airlines Inc.                    8.800352   -63.0  1553.0  
Frontier Airlines Inc.                     14.213311   -74.0  1253.0  
GoJet Airlines, LLC d/b/a United Express    6.325496  -475.0  1543.0  
Hawaiian Airlines Inc.                      0.851310   -67.0  2475.0  
Horizon Air                                 1.693461   -73.0   642.0  
JetBlue Airways                            11.432516   -79.0  1473.0  
Mesa Airlines Inc.                          9.088688   -63.0  1773.0  
Peninsula Airways Inc.                      7.485307   -50.0   437.0  
Republic Airlines                           3.478926   -77.0  1346.0  
SkyWest Airlines Inc.               

In [22]:
df['DepTime2'] = df['DepTime'] / 60

In [23]:
df.head()

FlightDate            Airline Origin Dest  Cancelled  Diverted  CRSDepTime  \
0  2018-01-23  Endeavor Air Inc.    ABY  ATL      False     False        1202   
1  2018-01-24  Endeavor Air Inc.    ABY  ATL      False     False        1202   
2  2018-01-25  Endeavor Air Inc.    ABY  ATL      False     False        1202   
3  2018-01-26  Endeavor Air Inc.    ABY  ATL      False     False        1202   
4  2018-01-27  Endeavor Air Inc.    ABY  ATL      False     False        1400   

   DepTime  DepDelayMinutes  DepDelay  ArrTime  ArrDelayMinutes  AirTime  \
0   1157.0              0.0      -5.0   1256.0              0.0     38.0   
1   1157.0              0.0      -5.0   1258.0              0.0     36.0   
2   1153.0              0.0      -9.0   1302.0              0.0     40.0   
3   1150.0              0.0     -12.0   1253.0              0.0     35.0   
4   1355.0              0.0      -5.0   1459.0              0.0     36.0   

   CRSElapsedTime  ActualElapsedTime  Distance  Year  Quarter  Month  \
0            62.0               59.0     145.0  2018        1      1   
1            62.0               61.0     145.0  2018        1      1   
2            62.0               69.0     145.0  2018        1      1   
3            62.0               63.0     145.0  2018        1      1   
4            60.0               64.0     145.0  2018        1      1   

   DayofMonth  DayOfWeek Marketing_Airline_Network  \
0          23          2                        DL   
1          24          3                        DL   
2          25          4                        DL   
3          26          5                        DL   
4          27          6                        DL   

  Operated_or_Branded_Code_Share_Partners  DOT_ID_Marketing_Airline  \
0                            DL_CODESHARE                     19790   
1                            DL_CODESHARE                     19790   
2                            DL_CODESHARE                     19790   
3                            DL_CODESHARE                     19790   
4                            DL_CODESHARE                     19790   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          DL                             3298   
1                          DL                             3298   
2                          DL                             3298   
3                          DL                             3298   
4                          DL                             3298   

  Operating_Airline  DOT_ID_Operating_Airline IATA_Code_Operating_Airline  \
0                9E                     20363                          9E   
1                9E                     20363                          9E   
2                9E                     20363                          9E   
3                9E                     20363                          9E   
4                9E                     20363                          9E   

  Tail_Number  Flight_Number_Operating_Airline  OriginAirportID  \
0      N8928A                             3298            10146   
1      N800AY                             3298            10146   
2      N8836A                             3298            10146   
3      N800AY                             3298            10146   
4      N8839E                             3298            10146   

   OriginAirportSeqID  OriginCityMarketID OriginCityName OriginState  \
0             1014602               30146     Albany, GA          GA   
1             1014602               30146     Albany, GA          GA   
2             1014602               30146     Albany, GA          GA   
3             1014602               30146     Albany, GA          GA   
4             1014602               30146     Albany, GA          GA   

   OriginStateFips OriginStateName  OriginWac  DestAirportID  \
0               13         Georgia         34          10397   
1               13         Georgia         34     

In [24]:
df.assign(DepTime3=df['DepTime'] / 60)

FlightDate                      Airline Origin Dest  Cancelled  \
0        2018-01-23            Endeavor Air Inc.    ABY  ATL      False   
1        2018-01-24            Endeavor Air Inc.    ABY  ATL      False   
2        2018-01-25            Endeavor Air Inc.    ABY  ATL      False   
3        2018-01-26            Endeavor Air Inc.    ABY  ATL      False   
4        2018-01-27            Endeavor Air Inc.    ABY  ATL      False   
...             ...                          ...    ...  ...        ...   
5689507  2018-09-11  Air Wisconsin Airlines Corp    SCE  IAD      False   
5689508  2018-09-11  Air Wisconsin Airlines Corp    IAD  GSO      False   
5689509  2018-09-11  Air Wisconsin Airlines Corp    EVV  ORD      False   
5689510  2018-09-11  Air Wisconsin Airlines Corp    ORD  HPN      False   
5689511  2018-09-11  Air Wisconsin Airlines Corp    HPN  ORD      False   

         Diverted  CRSDepTime  DepTime  DepDelayMinutes  DepDelay  ArrTime  \
0           False        1202   1157.0              0.0      -5.0   1256.0   
1           False        1202   1157.0              0.0      -5.0   1258.0   
2           False        1202   1153.0              0.0      -9.0   1302.0   
3           False        1202   1150.0              0.0     -12.0   1253.0   
4           False        1400   1355.0              0.0      -5.0   1459.0   
...           ...         ...      ...              ...       ...      ...   
5689507     False        1445   1433.0              0.0     -12.0   1515.0   
5689508     False        1235   1224.0              0.0     -11.0   1344.0   
5689509     False        1030   1016.0              0.0     -14.0   1137.0   
5689510     False        1410   1403.0              0.0      -7.0   1717.0   
5689511     False        1800   1754.0              0.0      -6.0   1919.0   

         ArrDelayMinutes  AirTime  CRSElapsedTime  ActualElapsedTime  \
0                    0.0     38.0            62.0               59.0   
1                    0.0     36.0            62.0               61.0   
2                    0.0     40.0            62.0               69.0   
3                    0.0     35.0            62.0               63.0   
4                    0.0     36.0            60.0               64.0   
...                  ...      ...             ...                ...   
5689507              0.0     35.0            61.0               42.0   
5689508              0.0     43.0            80.0               80.0   
5689509              0.0     54.0            94.0               81.0   
5689510              0.0    104.0           136.0              134.0   
5689511              0.0    116.0           153.0              145.0   

         Distance  Year  Quarter  Month  DayofMonth  DayOfWeek  \
0           145.0  2018        1      1          23          2   
1           145.0  2018        1      1          24          3   
2           145.0  2018        1      1          25          4   
3           145.0  2018        1      1          26          5   
4           145.0  2018        1      1          27          6   
...           ...   ...      ...    ...         ...        ...   
5689507     133.0  2018        3      9          11          2   
5689508     239.0  2018        3      9          11          2   
5689509     272.0  2018        3      9          11          2   
5689510     738.0  2018        3      9          11          2   
5689511     738.0  2018        3      9          11          2   

        Marketing_Airline_Network Operated_or_Branded_Code_Share_Partners  \
0                              DL                            DL_CODESHARE   
1                              DL                            DL_CODESHARE   
2                              DL                            DL_CODESHARE   
3                              DL                            DL_CODESHARE   
4                              DL                            DL_CODESHARE   
...                           ...                  

In [25]:
df.sort_values('DepTime', ascending=False)
#False means in the reverse order

FlightDate                      Airline Origin Dest  Cancelled  \
4643191  2018-08-21       Frontier Airlines Inc.    SEA  AUS      False   
4610100  2018-08-01            Republic Airlines    JFK  DCA      False   
3101244  2018-04-19         Delta Air Lines Inc.    ATL  RSW      False   
3953991  2018-06-25       Frontier Airlines Inc.    DEN  SEA      False   
4962771  2018-08-03       Southwest Airlines Co.    LGA  TPA      False   
...             ...                          ...    ...  ...        ...   
5688477  2018-09-15        United Air Lines Inc.    ORD  RDU       True   
5688583  2018-09-15        United Air Lines Inc.    MYR  EWR       True   
5688724  2018-09-12  Air Wisconsin Airlines Corp    ILM  IAD       True   
5688786  2018-09-12  Air Wisconsin Airlines Corp    IAD  FAY       True   
5688792  2018-09-12  Air Wisconsin Airlines Corp    ORD  ILM       True   

         Diverted  CRSDepTime  DepTime  DepDelayMinutes  DepDelay  ArrTime  \
4643191     False        2345   2400.0             15.0      15.0    613.0   
4610100     False        2210   2400.0            110.0     110.0    134.0   
3101244     False        2242   2400.0             78.0      78.0    124.0   
3953991     False        2205   2400.0            115.0     115.0    152.0   
4962771     False        1825   2400.0            335.0     335.0    231.0   
...           ...         ...      ...              ...       ...      ...   
5688477     False        2130      NaN              NaN       NaN      NaN   
5688583     False        1400      NaN              NaN       NaN      NaN   
5688724     False        1945      NaN              NaN       NaN      NaN   
5688786     False        1735      NaN              NaN       NaN      NaN   
5688792     False        1800      NaN              NaN       NaN      NaN   

         ArrDelayMinutes  AirTime  CRSElapsedTime  ActualElapsedTime  \
4643191             28.0    214.0           240.0              253.0   
4610100            106.0     54.0            98.0               94.0   
3101244             59.0     70.0           103.0               84.0   
3953991            110.0    153.0           177.0              172.0   
4962771            306.0    137.0           180.0              151.0   
...                  ...      ...             ...                ...   
5688477              NaN      NaN           117.0                NaN   
5688583              NaN      NaN           106.0                NaN   
5688724              NaN      NaN            73.0                NaN   
5688786              NaN      NaN            83.0                NaN   
5688792              NaN      NaN           136.0                NaN   

         Distance  Year  Quarter  Month  DayofMonth  DayOfWeek  \
4643191    1770.0  2018        3      8          21          2   
4610100     213.0  2018        3      8           1          3   
3101244     515.0  2018        2      4          19          4   
3953991    1024.0  2018        2      6          25          1   
4962771    1010.0  2018        3      8           3          5   
...           ...   ...      ...    ...         ...        ...   
5688477     646.0  2018        3      9          15          6   
5688583     550.0  2018        3      9          15          6   
5688724     323.0  2018        3      9          12          3   
5688786     284.0  2018        3      9          12          3   
5688792     760.0  2018        3      9          12          3   

        Marketing_Airline_Network Operated_or_Branded_Code_Share_Partners  \
4643191                        F9                                      F9   
4610100                        AA                            AA_CODESHARE   
3101244                        DL                                      DL   
3953991                        F9                                      F9   
4962771                        WN                                      WN   
...                           ...                  

In [26]:
df.isna().sum

<bound method DataFrame.sum of          FlightDate  Airline  Origin   Dest  Cancelled  Diverted  CRSDepTime  \
0             False    False   False  False      False     False       False   
1             False    False   False  False      False     False       False   
2             False    False   False  False      False     False       False   
3             False    False   False  False      False     False       False   
4             False    False   False  False      False     False       False   
...             ...      ...     ...    ...        ...       ...         ...   
5689507       False    False   False  False      False     False       False   
5689508       False    False   False  False      False     False       False   
5689509       False    False   False  False      False     False       False   
5689510       False    False   False  False      False     False       False   
5689511       False    False   False  False      False     False       False   

        

In [27]:
df[['FlightDate', 'Airline', 'ArrDelay']].dropna(subset=['ArrDelay'])

#It will only drop the rows where missing values occur in the subset(ArrDelay) column

FlightDate                      Airline  ArrDelay
0        2018-01-23            Endeavor Air Inc.      -8.0
1        2018-01-24            Endeavor Air Inc.      -6.0
2        2018-01-25            Endeavor Air Inc.      -2.0
3        2018-01-26            Endeavor Air Inc.     -11.0
4        2018-01-27            Endeavor Air Inc.      -1.0
...             ...                          ...       ...
5689507  2018-09-11  Air Wisconsin Airlines Corp     -31.0
5689508  2018-09-11  Air Wisconsin Airlines Corp     -11.0
5689509  2018-09-11  Air Wisconsin Airlines Corp     -27.0
5689510  2018-09-11  Air Wisconsin Airlines Corp      -9.0
5689511  2018-09-11  Air Wisconsin Airlines Corp     -14.0

[5586619 rows x 3 columns]

In [28]:
df['ArrDelay'].fillna(df['ArrDelay'].mean())
#Fills with the average value for better results

0          -8.0
1          -6.0
2          -2.0
3         -11.0
4          -1.0
           ... 
5689507   -31.0
5689508   -11.0
5689509   -27.0
5689510    -9.0
5689511   -14.0
Name: ArrDelay, Length: 5689512, dtype: float64

In [29]:
%pip install scikit-learn

Note: you may need to restart the kernel to use updated packages.


In [30]:
pip install --upgrade pip

Note: you may need to restart the kernel to use updated packages.


In [31]:
import sklearn
print(sklearn.__version__)

1.7.1


In [41]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix


In [33]:
y = df['DepDel15']

In [34]:
X = df[['CRSDepTime', 'CRSElapsedTime', 'Month', 'DayofMonth', 'DayOfWeek', 'Quarter',
        'Origin', 'Dest', 'Distance',
        'Marketing_Airline_Network', 'Operating_Airline',
        'Cancelled', 'Diverted', 'TaxiOut', 'DepTimeBlk']]

In [35]:
X = pd.get_dummies(X, drop_first=True)

In [36]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [37]:
df = df.sample(20000, random_state=42)

In [38]:
print(df.shape)  # rows, columns
print(df.memory_usage(deep=True).sum() / 1024**2, "MB")

(20000, 62)
25.705018997192383 MB


In [52]:
df = df.dropna(subset=['DepDel15'])

# Define target and features
y = df['DepDel15']
X = df.drop(columns=['DepDel15'])

# Encode categorical variables
cat_cols = X.select_dtypes(include='object').columns
le = LabelEncoder()
for col in cat_cols:
    X[col] = le.fit_transform(X[col].astype(str))

# Fill remaining NaNs
X = X.fillna(0)

# Split into train/test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

# Train model with class balancing
log_model = LogisticRegression(max_iter=500, solver='saga', class_weight='balanced')
log_model.fit(X_train, y_train)

# Predict probabilities
y_probs = log_model.predict_proba(X_test)[:, 1]

import numpy as np
from sklearn.metrics import precision_recall_curve, confusion_matrix, classification_report

# Compute precision-recall curve
precisions, recalls, thresholds = precision_recall_curve(y_test, y_probs)

# Find threshold that balances precision and recall
best_threshold = thresholds[np.abs(precisions - recalls).argmin()]
print("Best Threshold:", best_threshold)

# Apply optimal threshold
y_pred_opt = (y_probs > best_threshold).astype(int)

# Evaluate
print("Confusion Matrix (Optimal Threshold):")
print(confusion_matrix(y_test, y_pred_opt))

print("\nClassification Report (Optimal Threshold):")
print(classification_report(y_test, y_pred_opt))

Best Threshold: 0.5624306156035352
Confusion Matrix (Optimal Threshold):
[[4007  784]
 [ 785  330]]

Classification Report (Optimal Threshold):
              precision    recall  f1-score   support

         0.0       0.84      0.84      0.84      4791
         1.0       0.30      0.30      0.30      1115

    accuracy                           0.73      5906
   macro avg       0.57      0.57      0.57      5906
weighted avg       0.73      0.73      0.73      5906



/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


The Logistic Regression model achieved an overall accuracy of 73% after threshold tuning, with balanced performance between predicting on-time and delayed flights. However, the model struggled with delays specifically, achieving only 30% precision and 30% recall for the delayed class. This indicates that while the model can reasonably predict on-time flights, it is less effective at identifying delays—likely due to class imbalance, lack of feature scaling, and limited feature engineering.

Future improvements could include:
	•	Feature scaling for better convergence.
	•	Handling class imbalance (e.g., oversampling delays or using SMOTE).
	•	Using more advanced models (Random Forest or Gradient Boosting) that typically perform better on structured datasets like flight data.
